### ISSUE TITLE
Turn this repo into a website and support code sample in other langueges

### ISSUE THREAD

**Issue by letientai299** [kamranahmedse/design-patterns-for-humans#10]
Personally, I prefer to see Java or C++ rather than PHP. And, I don't want to violate DRY by duplicate the README to support other languages.

How do you think about a github.io page?

**Comment by Cybuster** [comment]
I was going to suggest the same, +1 for C++.

**Comment by Pahitos** [comment]
Yeah, I'd love to see this in Swift. 
But Java at the very least would be vice too. PHP is kind of readable I guess, but it takes way longer, since I'm not too familiar with it personally.

**Comment by aframe82** [comment]
OR, at least have a different branch for each language.

And yes, +1!

**Comment by tiagocesar** [comment]
Fork, work, pull request. No need to ask for it if you can produce it yourself.

**Comment by nloveladyallen** [comment]
@tiagocesar There's no reason to start working on it without reason to believe it will be merged.

**Comment by mstuefer** [comment]
@kamranahmedse please have a look at https://mstuefer.github.io/design-patterns-for-humans/ , I just started what could get a PR to add gh-pages. So far I only created the general-introduction-, the general-constructional-pattern and the SimpleFactory-page, lemme know what you think about. If you like it, we could take the information from the readme in a .json and create a mustache template with which we fill with the json etc.

**Comment by kamranahmedse** [comment]
Hi @mstuefer thank you for the effort but I have already started it and expect to have it ready by the next weekend or so. Forgot to add the in progress tag. Thanks for the offer to help though ☺

**Comment by kamranahmedse** [comment]
Plus, I have something else in mind for this. The design patterns will be split into hierarchical directory structure and there will be a command that will prepare the website. So that it is easier to modify and add more design patterns later on.

**Comment by mstuefer** [comment]
@kamranahmedse sounds good! Regarding the command, that is what I meant with the mustache template example, s.t. the usage of mustache & .json, e.g. via gulp.js, makes it easier to add/modify stuff etc.

**Comment by s2t2** [comment]
I was looking for ruby. Thanks.

**Comment by robkorv** [comment]
In what language are the current examples in?

**Comment by joereynolds** [comment]
@robkorv PHP

**Comment by segmentationfaulter** [comment]
Nobody asked for Javascript :disappointed:

**Comment by Crystalds93** [comment]
Nobody asked for JavaScript

**Comment by UsulPro** [comment]
+1 for Javascript :) I'm ready to join if someone'll start it

**Comment by aloisdg** [comment]
I started a [fork for C#](https://github.com/aloisdg/design-patterns-for-humans/).

I think I will made one for python too.

**Comment by nertus** [comment]
Nice to collect all languages versions in one online book

**Comment by nertus** [comment]
As optional - pseudocode is appropriate for all developers.

**Comment by aloisdg** [comment]
@nertus You miswrite Python ;)

**Comment by crane-yuan** [comment]
@kamranahmedse  I will translate this project into Chinese. :smile:
For now, I will use my more familiar with the Java language to rewrite the code below.

**Comment by Russell1227** [comment]
10

**Comment by DiglidiDudeNG** [comment]
The thing with providing examples in Javascript is that there are so many different ways to implement OOP as we know it in other more strongly-structured languages that it's better to base yourself around how it's implemented in a strong-typed, more structured language like C++ or Java and implement it in the way you're most comfortable with.

What I mean is, in JS, there are many different ways to implement the functionalities of OOP like classes, abstraction, instances, polymorphism, encapsulation etc. depending on the context. Either you can do it with ES6 `class` structure, or you can modify prototypes, or you can do it in an AMD or CommonJS module. It all depends on the context, so making it in Javascript would be more trouble than it's worth, unless you expect to have 3+ examples of javascript OOP implementations per design pattern.

**TL;DR**: Common OOP examples from other languages would suffice; making some for JS would be either time-consuming to write examples for every common contexts, or it'd be biased for or against certain OOP implementations that are still valid.

**Comment by letientai299** [comment]
@DiglidiDudeNG I think it's just a matter of code segment label (`JavaScript-ES6`, `JavaScript-ES5`, ...). People will have their own judgment for the usage.

**Comment by g3r4n** [comment]
@kamranahmedse a state of progress on this one ?

**Comment by mirkods** [comment]
👍  for Javascript/Nodejs :-)

**Comment by VictorAvelar** [comment]
@kamranahmedse  I can help with the website design if you need so.

**Comment by ryanfwy** [comment]
g

**Comment by Alienpruts** [comment]
After reading all this, I'm not too sure that my PR #66 is at all relevant. However, my line of thinking was that people tend to follow along with tutorials whilst 'debugging' code in their IDE. So in that respect I started this PR so all code examples could be 'translated' to other languages and made available for people to dig in using their IDE. I've already provided some examples in PHP, but the directory structure allows for more languages to be added.

Not too sure if that is still relevant though, seeing that quite some work has already been put in providing a online 'book'. :)

**Comment by yuduowu** [comment]
Hi guys, I've created **C++** and **Python** examples for my own learning purposes. You can find it here: https://github.com/yuduowu/design-patterns-for-humans/tree/master/design_patterns. I'm not C++/Python expert, some classes/functions may not be the best practices. Welcome any suggestions and discussions!

**Comment by Alienpruts** [comment]
Very nice @yuduowu 

If #66 gets in, maybe you can also do a pull request , I've already implemented a folder structure so that other languages can be inserted as examples.

Thanks again for your work :)

**Comment by mithunrt** [comment]
I'd love to see the code samples in JavaScript (ECMAScript 6+).

**Comment by ChucKN0risK** [comment]
I totally agree with you @mithunrt. This repo picked my interest because I thought it dealt with JavaScript.

**Comment by metodribic** [comment]
It would be best fit for all if this converts to github page where you could choose preferred language. I believe we could provide wide range of languages very quickly.

**Comment by phastmike** [comment]
I've created some Vala examples, which compile and can be 'executed':
[vala-design-patterns-for-humans](https://github.com/phastmike/vala-design-patterns-for-humans)

**Comment by Egorvoz** [comment]
Chat maybe we can discuss how the website will look like ?
We can use free hosting.
For example for good design we can use [ Bulma.io ](http://bulma.io/) CSS Framework (via Flexbox).

It will be good to create sections with languages and sections with different languages (examples) or only sections with different translations and every example will be in any programming language (some switcher) ? 

Also we can use Github Gist to highlight our code ? Or use JSFiddle or any kind of tools.

**Comment by kamranahmedse** [comment]
@Egorvoz website looks like below. It will be hosted on Github pages, is being built using React JS and Bootstrap 4

![image](https://cloud.githubusercontent.com/assets/4921183/23889499/4a8bf9c6-08af-11e7-981b-1c80540afc15.png)

**Comment by Egorvoz** [comment]
@kamranahmedse Awesome. Thanks.

**Comment by VictorAvelar** [comment]
@kamranahmedse awesome!

**Comment by IAmSurajBobade** [comment]
Interesting... Waiting for the launch!

Regards:
   *Suraj Ajinath Bobade*,
   *Computer Science Major at WCE, Sangli *
   suraj.bobade@outlook.com | *iamsurajbobade@gmail.com
<iamsurajbobade@gmail.com>*
   +91-7588 020 439

**Comment by bighead1226** [comment]
Great

**Comment by bighead1226** [comment]
Great
